src/main.ts

```diff
@@ -0,0 +1,656 @@
+
+  const secure = c.req.url.startsWith("https://");
+
+  const value = pipe(state, JSON.stringify, stringToBase64);
+  setCookie(c, "state", value, {
+    maxAge: 10 * seconds.minute,
+    httpOnly: true,
+    ...(secure ? { secure: secure, sameSite: "none" } : {}),
+  });
+  const valueUser = pipe(userID, JSON.stringify, stringToBase64);
+  setCookie(c, "user", valueUser, {
+    maxAge: 10 * seconds.minute,
+    httpOnly: true,
+    ...(secure ? { secure: secure, sameSite: "none" } : {}),
+  });
+
+  return c.redirect(url.toString());
+});
+
+app.get("/auth/github/callback", async (c) => {
+  const cookie = getCookie(c, "state");
+  const userID = getCookie(c, "user");
+
+  if (!userID) {
+    return c.json({ message: "unauthorized" }, { status: 401 });
+  }
+
+  if (!cookie) {
+    return c.json({ message: "No state" }, { status: 400 });
+  }
+
+  const state = pipe(cookie, base64ToString, JSON.parse);
+  const response = await authz.github.callback(
+    new URL(c.req.url),
+    state,
+    GITHUB_CALLBACK.toString()
+  );
+
+  const currentTimestamp = Date.now();
+
+  const endDate =
+    currentTimestamp + (response.refresh_token_expires_in as number) * 1000;
+
+  const octokit = new Octokit({ auth: response.access_token });
+  const userName = (await octokit.request("GET /user")).data.login;
+
+  // encrypt the refresh token
+  const bytes = await encrypt(
+    new TextEncoder().encode(response.refresh_token),
+    key
+  );
+  const buffer = Buffer.from(bytes);
+  const base64String = buffer.toString("base64");
+
+  // Save userId, refresh token, expiration date of the refresh token and the user name to db
+  db.query(
+    "insert into tokens (user_id, refresh_token, expiration_date, user_name) values ($userId, $refreshToken, $endDate, $userName)"
+  ).all({
+    $userId: userID!,
+    $refreshToken: base64String,
+    $endDate: endDate,
+    $userName: userName,
+  });
+
+  // save the userId, access token and the expiration date of the access token in memory
+  const endDate_access =
+    currentTimestamp + (response.expires_in as number) * 1000;
+  map.set(userID, {
+    accessToken: response.access_token,
+    expiresEpoch: endDate_access,
+  });
+  return c.redirect(`/?user_id=${userID}`);
+});
+
+app.get("tokens", async (c) => {
+  const userID = c.req.query("user_id");
+
+  if (!userID) {
+    return c.json("Unauthorized", 401);
+  }
+
+  const rows = await db
+    .query("select * from tokens where user_id = $userID")
+    .all({ $userID: userID });
+
+  return c.json(rows);
+});
+
+app.get("commits", async (c) => {
+  const userID = c.req.query("user_id");
+
+  if (!userID) {
+    return c.json({ message: "unauthorized" }, { status: 401 });
+  }
+
+  const rows = (await db
+    .query("select * from tokens where user_id = $userID")
+    .all({ $userID: userID as string })) as TokenRow[];
+
+  if (rows.length === 0) {
+    throw new Error("User not found in database. Log in with github first");
```

Comment:
in worst case this will crash the server, if hono has an error handling by default, it will return 500, which is incorrect status code

should probably be 404  ( loog at the if (!userID) to see an example)